helix-term/src/commands.rs

```diff
@@ -3284,33 +3286,88 @@ fn select_register(cx: &mut Context) {
     })
 }
 
-fn space_mode(cx: &mut Context) {
-    cx.on_next_key(move |cx, event| {
-        if let KeyEvent {
-            code: KeyCode::Char(ch),
-            ..
-        } = event
-        {
-            // TODO: temporarily show SPC in the mode list
-            match ch {
-                'f' => file_picker(cx),
-                'b' => buffer_picker(cx),
-                's' => symbol_picker(cx),
-                'w' => window_mode(cx),
-                'y' => yank_joined_to_clipboard(cx),
-                'Y' => yank_main_selection_to_clipboard(cx),
-                'p' => paste_clipboard_after(cx),
-                'P' => paste_clipboard_before(cx),
-                'R' => replace_selections_with_clipboard(cx),
-                // ' ' => toggle_alternate_buffer(cx),
-                // TODO: temporary since space mode took its old key
-                ' ' => keep_primary_selection(cx),
-                _ => (),
-            }
+macro_rules! mode_info {
+    // TODO: how to use one expr for both pat and expr?
+    // TODO: how to use replaced function name as str at compile time?
+    // TODO: extend to support multiple keys, but first solve the other two
+    {$name:literal, $cx:expr, $($key:expr => $func:expr; $funcs:literal),+,} => {
+        mode_info! {
+            $name, $cx,
+            $($key; $key => $func; $funcs,)+
         }
-    })
+    };
+    {$name:literal, $cx:expr, $($key:expr; $keyp:pat => $func:expr; $funcs:literal),+,} => {
+        $cx.editor.autoinfo = Some(Info::key(
+            $name,
+            vec![
+                $(
+                (vec![$key], $funcs),
+                )+
+            ],
+        ));
+        $cx.on_next_key(move |cx, event| {
+            match event {
+                $(
+                $keyp => $func(cx),
+                )+
+                _ => {}
+            }
+        })
+    }
 }
 
+fn space_mode(cx: &mut Context) {
+    mode_info! {
+        "space mode", cx,
+        key!('f'); key!('f') => file_picker; "file picker",
+        key!('b'); key!('b') => buffer_picker; "buffer picker",
+        key!('s'); key!('s') => symbol_picker; "symbol picker",
+        key!('w'); key!('w') => window_mode; "window mode",
+        key!('y'); key!('y') => yank_joined_to_clipboard; "yank joined to clipboard",
+        key!('Y'); key!('Y') => yank_main_selection_to_clipboard; "yank main selection to clipboard",
+        key!('p'); key!('p') => paste_clipboard_after; "paste clipboard after",
+        key!('P'); key!('P') => paste_clipboard_before; "paste clipboard before",
+        key!('R'); key!('R') => replace_selections_with_clipboard; "replace selections with clipboard",
+        key!(' '); key!(' ') => keep_primary_selection; "keep primary selection",
```

Comment:
Even on second try I still can't get this correct or happy with it. I wish it could be more ergonomic but not sure how to do it without proc macro:
```rust
key!(' ') => keep_primary_selection,
```

Reply:
It's fine as is, we may use descriptions different from the function name that way

Reply:
Ah so you mean for `ctrl-i` and `tab`?

Reply:
I don't understand, I mean this
![image](https://user-images.githubusercontent.com/3809077/123804297-181d7600-d8bb-11eb-8ff7-24c4c93278d1.png)
can be different from the function name that way

Reply:
Yes, it can, but it's easier if we make it the same name as function. No?

Reply:
I thought it was meant to be a description rather than a name (functions name often describes the action so it's not totally different anyway). I think it's fine to provide a description string, and we won't be able to do what you want without a proc macro anyway

Reply:
Yeah I agree here, the name should not be the documentation.